docs/specifications/tests/Basic-TP/basic01.md

```diff
@@ -1,199 +1,360 @@
+         already listed as glue record or records.
+      4. For each IP address (glue and looked-up) add the IP address and
+         *Referral Zone Name* to the *Remaining Servers* set unless the IP
+         address is already listed in *Handled Servers*.
+   8. If exactly one of the following criteria is met then save the name server
+      IP address, zone name for which the server is name server (parent zone) and
+      the DNS response to the *SOA Child Query* into the *Parent Name Server IP*
+      set. Criteria:
+      * The DNS response is a [Referral] to *Child Zone* (owner name of the NS
+        records is *Child Zone*), or
+      * The DNS response is a [Referral] with at least one CNAME record in the
+        answer section where one CNAME record has *Child Zone* as owner name.
+      * The DNS response has the AA flag set.
+
+> *Note that the "parent zone name" is the name of the zone that the name server
+> in question is NS for (owner name of NS). The name of the name server is
+> irrelevant.*
+
+6. For each name server IP and parent zone ("Parent Zone") in the
+   *Parent Name Server IP* set, do the following steps, including for any name
+   servers added to the set by the steps below.
+   1. Send a [DNS Query] with query type NS and *Parent Zone* as query name to
+      the name server IP.
+   2. If no DNS response, then go to next name server IP.
+   3. If the [DNS Response] contains no NS records in the answer section with
+      *Parent Zone* as owner name, then go to next name server IP.
+   4. Extract the NS records from the answer section with *Parent Zone* as owner
+      name from [DNS Response].
+   5. For each NS record extract the name server name ("Name Server Name") in
+      the RDATA field and do:
+      1. Create a [DNS Query] with query type A, *Name Server Name* as query
+         name and RD flag set, and do a [DNS Lookup].
+      2. If the [DNS Response], if any, contains a list of A records (follow
+         any CNAME chain) in the answer section then create a set of the IPv4
+         addresses from the A records ("NS IPv4 Addresses").
+      3. Create a [DNS Query] with query type AAAA, *Name Server Name* as
+         query name and RD flag set, and do a [DNS lookup].
+      4. If the [DNS Response], if any, contains a list of AAAA records
+         (follow any CNAME chain) in the answer section then create a set of
+         the IPv6 addresses from the AAAA records ("NS IPv6 Addresses").
+      5. If the *NS IPv4 Addresses* set or the *NS IPv6 Addresses* set is
+         non-empty, then then for each IP address in the two sets do if the
+         address is not already listed in the *Parent Name Server IP* set:
+         1. [Send] *SOA Child Query* to the IP address.
+         2. If the [DNS Response], if any, meets exactly one of the
+            following two criteria then save the IP address, the parent
+            zone name and the DNS response to the *Parent Name Server IP* set.
```

Comment:
> We should really avoid updating a set inside a loop that traverses this very same set.

As a written description I think it works well, but when executing (even manually) you probably have create some temporary structure. I will think about it, and we can talk about it, but I will do no change in this iteration.

> And which zone name is it we're inserting here? It's _Parent Zone_, right?

Yes, I have clarified that.

> Because I guess the purpose of step 6 is to expand _Parent Name Server IP_ to include name servers that are only listed in the parent zone itself - not its delegation?

That is the purpose.

> In any case we also have _Child Zone_ to choose from. Maybe spell it out explicitly which zone should be used?

Yes.

Reply:
>> We should really avoid updating a set inside a loop that traverses this very same set. [...] Then you may choose to insert newly generated elements into either one of the Remaining or Handled sets, depending on what behavior you want. [...]
>
> As a written description I think it works well, but when executing (even manually) you probably have create some temporary structure. I will think about it, and we can talk about it, but I will do no change in this iteration.

We should at least clarify whether elements added after the starting the traversal should be included in or excluded from the traversal. Otherwise the have undefined behavior.

Reply:
> We should at least clarify whether elements added after starting the traversal should be included in or excluded from the traversal. Otherwise the have undefined behavior.

I agree, and I have already considered that, but maybe not clearly enough. It reads:
```
6. For each name server IP and parent zone ("Parent Server Address" and
   "Parent Zone", respectively) in the *Parent Information* set, do the following
   steps, including for any name servers added to the set by the steps below.
```

Do you want me to rewrite this to be more clear?

Reply:
No, that's good. Thanks for updating.